pydicom/valuerep.py

```diff
@@ -43,6 +43,312 @@
 # (the component separator '=' is handled separately)
 PN_DELIMS = {0x5e}
 
+# maximum allowed value length for string VRs
+# VRs with a maximum length of 2^32 (UR and UT) are not checked
+MAX_VALUE_LEN = {
+    "AE": 16,
+    "CS": 16,
+    "DS": 16,
+    "IS": 12,
+    "LO": 64,
+    "LT": 10240,
+    "SH": 16,
+    "ST": 1024,
+    "UI": 64
+}
+
+
+def _range_regex(regex: str) -> str:
+    """Compose a regex that allows ranges of the given regex,
+    as defined for VRs DA, DT and TM in PS 3.4, C.2.2.2.5.
+    """
+    return fr"^{regex}$|^\-{regex} ?$|^{regex}\- ?$|^{regex}\-{regex} ?$"
+
+
+# regular expressions to match valid values for some VRs
+AE_REGEX = re.compile(r"^[\x20-\x7e]*$")
+AS_REGEX = re.compile(r"^\d\d\d[DWMY]$")
+CS_REGEX = re.compile(r"^[A-Z0-9 _]*$")
+DS_REGEX = re.compile(
+    r"^ *[+\-]?(\d+|\d+\.\d*|\.\d+)([eE][+\-]?\d+)? *$"
+)
+IS_REGEX = re.compile(r"^ *[+\-]?\d+ *$")
+DA_REGEX = re.compile(_range_regex(r"\d{4}(0[1-9]|1[0-2])([0-2]\d|3[01])"))
+DT_REGEX = re.compile(_range_regex(
+    r"\d{4}((0[1-9]|1[0-2])(([0-2]\d|3[01])(([01]\d|2[0-3])"
+    r"([0-6]\d([0-5]\d(\.\d{1,6} ?)?)?)?)?)?)?([+-][01]\d\d\d)?")
+)
+TM_REGEX = re.compile(_range_regex(
+    r"([01]\d|2[0-3])([0-6]\d([0-5]\d(\.\d{1,6} ?)?)?)?"))
+UR_REGEX = re.compile(r"^[A-Za-z_\d:/?#\[\]@!$&'()*+,;=%\-.~]* *$")
+
+
+def validate_vr_length(vr: str, value: Union[str, bytes]) -> Tuple[bool, str]:
+    """Validate the value length for a given VR.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str or bytes
+        The value to validate.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    max_length = MAX_VALUE_LEN.get(vr, 0)
+    if max_length > 0:
+        value_length = len(value)
+        if value_length > max_length:
+            return False, (
+                f"The value length ({value_length}) exceeds the "
+                f"maximum length of {max_length} allowed for VR {vr}."
+            )
+    return True, ""
+
+
+def validate_regex(vr: str, value: Union[str, bytes],
+                   regex: Any) -> Tuple[bool, str]:
+    """Validate the value for a given VR for allowed characters
+    using a regular expression.
+
+    Parameters
+    ----------
+    vr : str
+        The value representation to validate against.
+    value : str
+        The value to validate.
+    regex : re.Pattern
+        The precompiled regex pattern used for matching.
+
+    Returns
+    -------
+        A tuple of a boolean validation result and the error message.
+    """
+    if not re.match(regex, value):
+        return False, f"Invalid value for VR {vr}: '{value!r}'."
+    return True, ""
+
+
+def validate_length_and_regex(vr: str, value: Union[str, bytes],
+                              regex: Any) -> Tuple[bool, str]:
```

Comment:
You should be able to use `regex: re.Pattern[str]`, I think.

Reply:
I used that, and it didn't work for something - I think Python 3.6 and/or Pypy, have to check.

Reply:
Yes, I got:
```
ERROR  - AttributeError: module 're' has no attribute 'Pattern'
```
both in 3.6 and (3.6-based) pypy

Reply:
Weird, I tested the following with 3.6.5:
```python
import re
from typing import Optional

def foo(r: re.Pattern[str], bar: str) -> Optional[re.Match[str]]:
    return re.match(r, bar)

if __name__ == "__main__":
    foo(re.compile(r"[a-z]"), "asdf")
```

Reply:
https://github.com/pydicom/pydicom/runs/4417100547?check_suite_focus=true
I had used `re.Pattern` instead of `re.Pattern[str]` at the time, but that doesn't change the outcome (just checked).